Add ability to compile with VCRT

am11 opened the pull request:
Addresses #2

@Snaipe, I have also added CI for MinGW (failing 4/5 tests) and VCRT (failing 5/5). Can you please take a look at appveyor.yml and advise if there is anything required to setup tests properly?

am11 commented on appveyor.yml:
```diff
@@ -0,0 +1,49 @@
+      -Wno-dev
+      -DCMAKE_INSTALL_PREFIX=boxfort-%RELEASE_NAME%
+      -DCMAKE_BUILD_TYPE="%CONFIGURATION%"
+      -DBXF_STATIC_LIB=ON ..
```
Without this the build fails as if compiling with shared lib is not possible. The linker looks for boxfort.lib even in case of shared build.

Snaipe replied:
Weird, I'll take a look. You also need to add the `build/Debug/` directory to the PATH environment variable for the shared lib to get discovered at runtime.

am11 replied:
I set the PATH, but apparently it is not looking for `boxfort.dll` but complaining about `boxfort.lib` https://ci.appveyor.com/project/am11/boxfort/build/1.0.0_b12-build/vcr.

So I reverted this line to explicitly build as static lib (fixed the typo; removed `..` from this line). Now the library is build: https://ci.appveyor.com/project/am11/boxfort/build/1.0.0_b13-build/vcr (failing unit tests though)

Maybe we need to have (a simplified version of) what's there in Criterion:
https://github.com/Snaipe/Criterion/blob/4d6d741/.cmake/Modules/PackageUtils.cmake#L28-L57

am11 commented on src/asm/trampoline-i386.asm:
```diff
@@ -0,0 +1,60 @@
+  and     eax, 0fffff000h
+  push    eax
+  mov     eax, dword ptr [eax]                      ; Setup mock context
+  push    eax                                       ; Call mmk_set_ctx
```
oops `mmk_set_ctx` 😄 
It was *mimicked* from Mimick, then adjusted the symbol names. Will fix after your review.

Snaipe replied:
Aha, this is a trick. The trampoline used by Mimick is a different kind that does much more than jumping, in the case of BoxFort the trampoline is a jump instruction that uses an address written right after the instruction itself. See trampoline-x86_64 for more details.

am11 replied:
Thanks for the pointers and yay! I finally am able to transform the AT&T syntax to Intel in: f7dcf07. Not sure if there's a bug there, but apparently assembler and linker are happy with the code bits. :)

am11 commented on src/asm/trampoline-x86_64.asm:
```diff
@@ -0,0 +1,58 @@
+  and     rax, 0fffff000h
+  push    rax
+  mov     rax, qword ptr [rax]                      ; Setup mock context
+  push    rax                                       ; Call mmk_set_ctx
```
same as above.

am11 commented on src/exe-pe.c:
```diff
@@ -57,7 +61,13 @@ int bxfi_exe_patch_main(bxfi_exe_fn *new_main)
+    char opcodes[size]; // VLA
+#else
+    char *opcodes = alloca(size);
+#endif
```
This was breaking the build on VCRT.

Snaipe commented on appveyor.yml:
```diff
@@ -0,0 +1,49 @@
+environment:
+  TERM: dumb
+  COVERALLS_REPO_TOKEN:
+    secure: 5nuCg+faxFPeppoNNcSwVobswAVFUf8ut83vw8CX/4W2y0kZkGmwEfCUxSQWiQDU
```
No need for a coveralls repo token here. I'll setup this in time.

Snaipe commented on appveyor.yml:
```diff
@@ -0,0 +1,49 @@
+
+configuration:
+  - Debug
+  - Release
```
I think that for faster overall builds I'm no longer going to build the Release config, just RelWithDebInfo & Debug. Usually the Release build does not provide any more insights than RelWithDebInfo.

am11 replied:
Removed the release build. And added the artifacts rule which zips the build directory for each job (and downloadable from API and UI, e.g. https://ci.appveyor.com/project/am11/boxfort/build/1.0.0_b18-build/vcr/job/57iy8w6cmr6oo174/artifacts)

Snaipe commented:
As a side note don't worry too much about passing the tests *yet*. There are a few things not working completely on windows that need some debugging, and this shouldn't be in the scope of this PR.

am11 commented:
Thanks. I have disabled the test run with `test: off` in the YAML file. We can remove that line once the unit tests start passing.

Snaipe commented:
As far as the issue with `boxfort.lib` not existing, this happens because BoxFort does not export its API functions. I'll roll a fix on master for that.

Snaipe commented on include/boxfort.h:
```diff
@@ -33,12 +33,14 @@
 #  ifdef BXF_BUILDING_LIB
 #   ifdef __GNUC__
 #    define BXF_API __attribute__((dllexport))
+#    define BUILDING_DLL
```
(fyi I forgot to define BXF_BUILDING_LIB from the cmake configuration, this is now fixed on master)

am11 replied:
Oh great, I was trying to fix MinGW failure, thanks :)

am11 commented:
I rebased against master, and now it is building VCRT win32 without static flag. ☑️ 

MinGW continues to pass after the rebase. ☑️ 

I have also added VCRT win64 jon in matrix and it is throwing linker error: ❎ 

> 198‍‍trampoline-x86_64.obj : error LNK2017: 'ADDR32' relocation to 'addr_data' invalid without /LARGEADDRESSAWARE:NO [C:\projects\boxfort\build\boxfort.vcxproj]
 199‍‍LINK : fatal error LNK1165: link failed because of fixup errors

Is there anything that can be fixed in trampoline-x86-64, without disabling `LARGEADDRESSAWARE` switch?

Snaipe commented:
Try using `qword ptr addr_data` directly, x86_64 introduces PC-relative addressing

am11 commented:
GREAT! That did the trick. 🎉 
I actually tried `qword ptr addr_data[rip]` but `rip` is not available in MASM, so I thought explicitly retrieving the IP by calling label is the way to go (as has been done in Mimick). Another approach I found was to move a label `mov rax, next`, then rax has the program counter. But `call next` was comparatively neat. :)

am11 commented:
Thanks! If you add BoxFort in AppVeyor, the badge in README will light up! 💡
